Index: ssh_config
===================================================================
--- ssh_config.orig
+++ ssh_config
@@ -63,6 +63,9 @@ ForwardX11Trusted yes
 # potential man-in-the-middle attacks, which 'gssapi-with-mic' is not susceptible to.
 #   GSSAPIEnableMITMAttack no
 
->>>>>>>
+# This enables sending locale enviroment variables LC_* LANG, see ssh_config(5).
+SendEnv LANG LC_CTYPE LC_NUMERIC LC_TIME LC_COLLATE LC_MONETARY LC_MESSAGES
+SendEnv LC_PAPER LC_NAME LC_ADDRESS LC_TELEPHONE LC_MEASUREMENT
+SendEnv LC_IDENTIFICATION LC_ALL
 #   VisualHostKey no
 #   ProxyCommand ssh -q -W %h:%p gateway.example.com
Index: sshd_config
===================================================================
--- sshd_config.orig
+++ sshd_config
@@ -117,6 +117,11 @@ X11Forwarding yes
 # override default of no subsystems
 Subsystem	sftp	/usr/libexec/sftp-server
 
+# This enables accepting locale enviroment variables LC_* LANG, see sshd_config(5).
+AcceptEnv LANG LC_CTYPE LC_NUMERIC LC_TIME LC_COLLATE LC_MONETARY LC_MESSAGES
+AcceptEnv LC_PAPER LC_NAME LC_ADDRESS LC_TELEPHONE LC_MEASUREMENT
+AcceptEnv LC_IDENTIFICATION LC_ALL
+
 # Example of overriding settings on a per-user basis
 #Match User anoncvs
 #	X11Forwarding no
